### ISSUE TITLE
Wrong result for bit-wise negating then widening a calculation

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#983]
There might be related problems that I haven't discovered yet, but I have found some issues where if you do bit-wise negation on a small bit vector, then assign it or use it in a calculation that casts it to a wider value, like this:

```
     bit<16> tmp = ~foo;      // e.g. suppose that foo is 16w0x0800, so tmp should become 16w0xf7ff
     bit<32> x32 = (bit<32>) tmp;   // x32 should become 32w0xf7ff, extending with 0 bits in MSBs
```
the generated JSON file for simple_switch is the same as that generated if you first cast the small bit vector to a wide one, then negate it.
```
    bit<32> y32 = ~((bit<32>) tmp);   // y32 should become 32w0xfffff7ff
```
with a recent version of p4c-bm2-ss and simple_switch, I see x32 become 32w0xfffff7ff, the same as y32 becomes.

I will attach a sample program here soon.

**Comment by jafingerhut** [comment]
[issue983.p4.txt](https://github.com/p4lang/p4c/files/1407875/issue983.p4.txt)

I may also create an STF file to exercise this program, but the basic idea is that if you send in a packet with just an Ethernet header with etherType 0x0800, if everything is working as I understand the P4_16 spec, it should output a packet with an ethernet dstAddress that is all 0, indicating no errors.  Currently it outputs a non-0 dstAddress, indicating several unexpected results for arithmetic calculations.

**Comment by jafingerhut** [comment]
I am not sure, but for one of the 'error codes' this program generates (the one for the last 'if' statement in the ingress control block), it might be due to a bug in simple_switch.

**Comment by antoninbas** [comment]
I'm guessing that's because bmv2 uses libgmp, which assumes arbitrary size.
When doing a one's complement, the result is extended with infinite 1s. For example, the complement of 0000...0011 is ...1111...1100. This means that a masking operation has to be performed right after the one's complement operation to guarantee correctness.

**Comment by jnfoster** [comment]
...but the compiler front-end passes are careful to insert the correct
coercions during type checking. So I'm not sure.

-N

**Comment by jafingerhut** [comment]
[issue983.zip](https://github.com/p4lang/p4c/files/1427789/issue983.zip)

The attached zip file contains the original sample program attached above, renamed to issue983-bmv2.p4 (with one small change for recent verifyChecksum in -> inout parameter direction change).  It also contains a small issue983-bmv2.stf file that currently fails, but should pass if the issue is fixed.

It also contains issue983-workaround-bmv2.p4, which is very similar to issue983-bmv2.p4, but contains some extra "& 0xfff" and "& 0xffff_ffff" in a few expressions.  There are probably other ways to do it, but with those extra operations the compiled JSON file with today's p4c and simple_switch give the expected results.

**Comment by jafingerhut** [comment]
[issue983b.zip](https://github.com/p4lang/p4c/files/1444612/issue983b.zip)

I started poking into the root cause of this bug a little bit, hoping that perhaps I might find a way to fix it.  I have attached a .zip file that contains a patch file that prints some extra debug output while running p4c-bm2-ss on a fairly minimal P4 program that exhibits the bug, also included in the zip file.  There is a README file in there, too, showing the extra debug output produced, and some comments on what that output seems to imply.  Fixing this bug looks beyond my fairly small knowledge of p4c internals, so I wanted to include what I had learned in case it helps speed up the job for someone else (or me if I learn more later).

**Comment by jafingerhut** [comment]
With a similar 'minimal' P4 program with this assignment, doing addition of `bit<16>` values instead of bitwise negation, but also then casting to `bit<32>` after the arithmetic operation, like this:
```
hdr.ethernet.dstAddr_lo = (bit<32>) (hdr.ethernet.srcAddr_lo[15:0] + hdr.ethernet.srcAddr_lo[31:16]);
```
the same patch that prints extra debug messages in my previous comment's attachment shows the assignment becomes the following as input to the bmv2 backend code:
```
hdr.ethernet.dstAddr_lo = (bit<32>) ((bit<16>)hdr.ethernet.srcAddr_lo[15:0] + (bit<16>)hdr.ethernet.srcAddr_lo[31:16]);
```
The extra `bit<16>` occurrences that the compiler added somewhere along the way is causing the bmv2 backend to generate correct '& 0xffff' operations on those `bit<16>` values, whereas with the bit-wise negation of a `bit<16>` value, the compiler is not inserting the 'extra' `bit<16>` cast on the operand.  If the compiler treated bit-wise negation more like it is already treating addition, at some compiler pass that I don't know what that would be yet where it adds these casts in the case of addition, then I suspect that would fix this bug.

**Comment by mbudiu-vmw** [comment]
It may be sufficient to add the complement to the ArithmeticFixup visitor in the BMv2 back-end.
This is the one that inserts additional casts.

**Comment by jafingerhut** [comment]
I've re-tested a larger program that led to this issue being reported, after the proposed fix in PR #1011 was committed, and behavior is better now.  Thanks!
